File: 516 5-smooth totients.pl
#!/usr/bin/perl

# Daniel "Trizen" Șuteu
# License: GPLv3
# Date: 04 May 2017
# https://github.com/trizen

# https://projecteuler.net/problem=516

# Not a very good solution: requires over 6 GB of RAM.

# Runtime: 2 min, 57 sec.

use 5.010;
use strict;

use ntheory qw(
  factor
  euler_phi
  is_square_free
  );

my @smooth = (1);
my $limit  = 1e12;
my $mod    = 1 << 32;

foreach my $p (2, 3, 5) {
    foreach my $n (@smooth) {
        if ($n * $p <= $limit) {
            push @smooth, $n * $p;
        }
    }
}

@smooth =
  sort { $b <=> $a }
  grep { is_square_free($_) and ((factor(euler_phi($_)))[-1] <= 5) }
  map { $_ + 1 } @smooth;

my @h = (1);
my %seen;
my $sum = 1;

foreach my $p (@smooth) {
    foreach my $n (@h) {
        if (
                $p * $n <= $limit
            and !exists($seen{$p * $n})
            and (factor(euler_phi($n * $p)))[-1] <= 5
        ) {

            if ($p == 2) {
                foreach my $n (@h) {
                    while (
                            $n * $p <= $limit
                        and !exists($seen{$p * $n})
                        and (factor(euler_phi($n * $p)))[-1] <= 5
                    ) {
                        $sum += ($n * $p) % $mod;
                        $sum %= $mod;
                        $n *= $p;
                    }
                }
                last;
            }
            else {
                undef $seen{$n * $p};
                $sum += ($n * $p) % $mod;
                $sum %= $mod;
                push @h, $n * $p;
            }
        }
    }

    say "$p -> $sum ($#h)";
}

say "Final answer: $sum";

File: 516 5-smooth totients.jl
#!/usr/bin/julia

# Daniel "Trizen" Șuteu
# License: GPLv3
# Date: 04 May 2017
# https://github.com/trizen

# https://projecteuler.net/problem=516

# Runtime: 1 min, 23 sec.

using Primes

function Φ(n::Int64)

    for p in keys(factor(n))
        n -= div(n, p)
    end

    n
end

function is_5_smooth(n::Int64)

    for p in [2,3,5]
        while (rem(n, p) == 0)
            n = div(n, p)
        end
    end

    n == 1
end

function p_516(limit::Int64 = 10^12)

    smooth = Int64[1]

    for p in [2,3,5]
        for n in smooth
            if (n*p <= limit)
                append!(smooth, n*p)
            end
        end
    end

    smooth = reverse(
        sort(
            filter((n)->isprime(n),
            map((n)->n+1, smooth))
        )
    )

    h   = Int64[1]
    mod = 1 << 32

    sum = 0
    for p in smooth
        for n in h
            if (p*n <= limit && is_5_smooth(Φ(n*p)))
                if (p == 2)
                    for n in h
                        while (n*p <= limit)
                            sum += (n*p) % mod
                            sum %= mod
                            n *= p
                        end
                    end
                    break
                else
                    sum += (n*p) % mod
                    sum %= mod
                    append!(h, n*p)
                end
            end
        end
        println("$p -> $sum")
    end

    return sum+1
end

println(p_516(10^12))
